src/persist-txn/src/txn_cache.rs

```diff
@@ -241,21 +240,16 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> TxnsCacheState
                 empty_to: self.progress_exclusive.clone(),
             };
         };
-        let lastest_forget = all
-            .registered
-            .iter()
-            .rev()
-            .find_map(|reg| reg.forget_ts.as_ref());
-        // Forgetting a shard guarantees that all previous writes have been applied.
-        let applied = |x| matches!(lastest_forget, Some(forget_ts) if forget_ts >= x);
-        // TODO(jkosh44) latest_write and empty_to are not right, `writes` can be compacted past
-        // an unapplied write, meaning we can miss a write.
-        // See https://github.com/MaterializeInc/materialize/issues/26893.
+        let min_unapplied_ts = self
+            .unapplied_batches
+            .first_key_value()
+            .map(|(_, (_, _, ts))| ts)
+            .unwrap_or(&self.progress_exclusive);
```

Comment:
I have to think about this a little more. Does it matter that we're not checking the shard ID?